mantidproject/mantid
# Fix `PDCalibration` with workspace indices

### Description of work

`PDCalibration` is largely a wrapper of `FitPeaks`.  Work was done in PR #35837  to expose the `StartWorkspaceIndex` and `StopWorkspaceIndex` properties of `FitPeaks` to `PDCalibration`.

`FitPeaks` also requires initial guesses and ranges for the peaks, which can be passed in one of two ways, as lists or as workspaces.  This was not written to work intuitively with the start/stop workspace index option, requiring some novel methods of matching indices.  The way that workspace indices were handled, and peak center/window workspaces were handled, led to some contradicting behavior.

According to the documentation, the peak center property should behave as follows:

> One of only one of the following will be taken: 
> 1. A MatrixWorkspace
>       -  Number of spectra shall be the same as the number of spectra of the workspace containing peaks to fit for. Or the number of spectra is the same as the number of spectra of the input workspace.
>        - X value is the index of the peak.
>        - Y value is the position of the peaks to fit.
>        - Peak centers are stored in m_peakCenterWorkspace.
>        - Spectra can be a subset of all spectra because FitPeaks can work on partial spectra.
>  2. An array of double as the positions of the peaks to fit.
>        - Peak centers are stored in m_peakCenters
> Peaks’ positions must be given in ascending order.

That the number of spectra shall be the same as the number of spectra of the input workspace, suggests the workspace indices in the peak center workspace should correspond to the same spectra as the workspace index in the input workspace.

As it was, even if `FitPeaks` was set to match workspace indices 2,3,4 (for example), then the peak center/window workspaces were still expected to define their lists at workspace indices 0,1,2.  Since those spectra have different peak locations in TOF, this meant the guesses pulled were bad, and peaks not fit.  This was causing an error inside `PDCalibration`.

This fixes the issue, so that the workspace indices for the peak centers and windows correspond to the workspace indices for the spectra in the input workspace.

### Purpose of work

In the diffraction calibration process used with SNAP at ORNL, it is desired to first diffraction focus the data, the run `PDCalibration` on one focuses spectrum at a time, as this allows exploiting the fact the `DIFC` value will be constant on all detectors within a focus group.  This requires being able to run `PDCalibration` a single spectrum at a time, which required prior work in PR #35837.  However, as `PDCalibration` passes the guesses for peak centers and windows as a workspace, that was causing errors with access by workspace index.

With `FitPeaks`, it should now be possible after this work to create one workspace, once, with peak center/window references matching all spectra in the input workspace, then fit peaks on each individual spectrum without having to build a new set of peak center/window input workspaces for each spectrum.

### Summary of work

When accessing peak centers or windows from a workspace, instead of something like
``` cpp
std::size_t peakwsindex = wsindex - m_startWorkspaceIndex;
const auto &peakCenterX = m_peakCenterWorkspace->x(peakwsindex);
```
(which was causing failures) this instead uses
``` cpp
const auto &peakCenterX = m_peakCenterWorkspace->x(wsindex);
```
where `wsindex` is the workspace index within the input WS and is between `m_startWorkspaceIndex` and `m_stopWorkspaceIndex`.

The output workspace is still treated the same, so that indices range from 0 to *N*, where *N* is the number of spectra.  Even if these spectra correspond to different workspace indices in the input workspace.

### Further detail of work

In addition to the changes mentioned above:

Inside `FitPeaks.cpp`:
- Added `m_numSpectraToFit` to avoid having to recalculate this quantity many times.
- Updated some documentation to better clarify the use of workspace indices.
- Updated some error messages to be more informative or clear.
- Fixed the way workspace indices are handled to require less casting between `size_t` and `int64`.
- Wherever it came up, changed `size_t` to `std::size_t`.
- Fixed `m_numPeaksToFit` to work with ragged peak center workspaces.
- **major refactor** the function `getExpectedPeakPositions` was called from inside a for loop over workspace indices.  This function used branching logic to determine how to find the expected peak positions, but the same method is always used for every spectrum.  Instead, this function has been replaced by a member `std::function` called `m_getExpectedPeakPositions` which is set once during the input processing, to avoid this branching.  Likewise with the function `getPeakFitWindow`.  This simplifies the flow of the program somewhat.

Inside `FitPeaksTest.h`
- Added loggers to identify the test, for easier debugging
- Set `FitPeaks` to rethrow errors, ensure nothing thrown at execution
- The test data workspaces use points rather than edges, so that the number of centers can be better compared to the number of window edges
- Added test to see what happens with a partial spectrum fit, when not including the zero spectrum in list to fit.  This test would have failed before the change, as it would have compared the data at spectrum 2 to the peak windows defined for spectrum 0.

### To test:

Check that the new unit tests run.

Check that docs build:
``` bash
cmake -DDCS_SCREENSHOTS=on .
ninja docs-html
```

Run the docs tests:
``` bash
ninja docs-doctest
```

Within the SNS analysis cluster, run the script that I will post in the comments below.  Check the fitted workspace of `PDCalibration`, overlapped with the TOF data, to ensure the peaks were fit.  They should be so sharp you need safety goggles.

Fixes #36123

Related to [EWM 2437](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20%28Change%20Management%29#action=com.ibm.team.workitem.viewWorkItem&id=2437) at ORNL


Release notes and improved documentation have been added to explain the workspace indices.

---

### Reviewer ###

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review ####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

#### Functional Tests ####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper ###

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
